src/stcal/resample/resample.py

```diff
@@ -0,0 +1,1506 @@
+
+__all__ = [
+    "OutputTooLargeError",
+    "ResampleModelIO",
+    "ResampleBase",
+    "ResampleCoAdd",
+    "ResampleSingle"
+]
+
+
+_SUPPORTED_CUSTOM_WCS_PARS = [
+    'pixel_scale_ratio',
+    'pixel_scale',
+    'output_shape',
+    'crpix',
+    'crval',
+    'rotation',
+]
+
+
+# FIXME: temporarily copied here to avoid this import:
+# from stdatamodels.jwst.library.basic_utils import bytes2human
+def bytes2human(n):
+    """Convert bytes to human-readable format
+
+    Taken from the `psutil` library which references
+    http://code.activestate.com/recipes/578019
+
+    Parameters
+    ----------
+    n : int
+        Number to convert
+
+    Returns
+    -------
+    readable : str
+        A string with units attached.
+
+    Examples
+    --------
+    >>> bytes2human(10000)
+        '9.8K'
+
+    >>> bytes2human(100001221)
+        '95.4M'
+    """
+    symbols = ('K', 'M', 'G', 'T', 'P', 'E', 'Z', 'Y')
+    prefix = {}
+    for i, s in enumerate(symbols):
+        prefix[s] = 1 << (i + 1) * 10
+    for s in reversed(symbols):
+        if n >= prefix[s]:
+            value = float(n) / prefix[s]
+            return '%.1f%s' % (value, s)
+    return "%sB" % n
+
+
+def _resample_range(data_shape, bbox=None):
+    # Find range of input pixels to resample:
+    if bbox is None:
+        xmin = ymin = 0
+        xmax = data_shape[1] - 1
+        ymax = data_shape[0] - 1
+    else:
+        ((x1, x2), (y1, y2)) = bbox
+        xmin = max(0, int(x1 + 0.5))
+        ymin = max(0, int(y1 + 0.5))
+        xmax = min(data_shape[1] - 1, int(x2 + 0.5))
+        ymax = min(data_shape[0] - 1, int(y2 + 0.5))
+
+    return xmin, xmax, ymin, ymax
+
+
+class ResampleModelIO(abc.ABC):
+    @abc.abstractmethod
+    def open_model(self, file_name):
+        ...
+
+    @abc.abstractmethod
+    def get_model_attr_value(self, model, attribute_name):
+        ...
+
+    @abc.abstractmethod
+    def set_model_attr_value(self, model, attribute_name, value):
+        ...
+
+    @abc.abstractmethod
+    def get_model_meta(self, model, attributes):
+        ...
+
+    @abc.abstractmethod
+    def set_model_meta(self, model, attributes):
+        ...
+
+    @abc.abstractmethod
+    def get_model_array(self, model, array_name):
+        ...
+
+    @abc.abstractmethod
+    def set_model_array(self, model, array_name, data):
+        ...
+
+    @abc.abstractmethod
+    def close_model(self, model):
+        ...
+
+    @abc.abstractmethod
+    def save_model(self, model):
+        ...
+
+    @abc.abstractmethod
+    def write_model(self, model, file_name, **kwargs):
+        ...
+
+    @abc.abstractmethod
+    def new_model(self, image_shape=None, file_name=None, copy_meta_from=None):
+        """ Return a new model for the resampled output """
+        ...
+
+
+class OutputTooLargeError(RuntimeError):
+    """Raised when the output is too large for in-memory instantiation"""
+
+
+def output_wcs_from_input_wcs(input_wcs_list, pixel_scale_ratio=1.0,
```

Comment:
I might put this somewhere other than `resample.py`, e.g. alignment/utils or alignment/resample_utils.  I would vote (eventually) for having a separate submodule for WCS-related utility functions, but that's probably beyond the PR scope

Reply:
Agree. It was temporary here. This will be fixed once we have a function that computes output WCS from `s_region` of input data models and other parameters.